rbmc-cfam-daemon/src/sibling_cfam.hpp

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: Apache-2.0
+#pragma once
+#include "bmc_cfam.hpp"
+
+#include <xyz/openbmc_project/State/BMC/Redundancy/common.hpp>
+#include <xyz/openbmc_project/State/BMC/common.hpp>
+
+#include <optional>
+
+/**
+ * @class SiblingCFAM
+ *
+ * Represents the CFAM on the sibling BMC's card.
+ *
+ * Handles reading fields from the CFAM.  The getter functions
+ * return the most recent value since the last readAll() call.
+ * If that call failed, they will all throw an exception until
+ * there is a good read again.
+ */
+class SiblingCFAM : public BMCCFAM
+{
+  public:
+    using Role =
+        sdbusplus::common::xyz::openbmc_project::state::bmc::Redundancy::Role;
+    using BMCState =
+        sdbusplus::common::xyz::openbmc_project::state::BMC::BMCState;
+
+    SiblingCFAM() = delete;
+    ~SiblingCFAM() override = default;
+    SiblingCFAM(const SiblingCFAM&) = delete;
+    SiblingCFAM& operator=(const SiblingCFAM&) = delete;
+    SiblingCFAM(SiblingCFAM&&) = delete;
+    SiblingCFAM& operator=(SiblingCFAM&&) = delete;
+
+    /**
+     * @brief Constructor
+     *
+     * @param[in] link - The link the CFAM is on
+     * @param[in] sysfs - The sysfs object
+     */
+    SiblingCFAM(size_t link, SysFS& sysfs) :
+        BMCCFAM(link, sysfs), usedRegs(BMCCFAM::getUsedRegs())
+    {}
+
+    /**
+     * @brief Returns the API version field
+     *
+     * Will throw if there is a hardware error
```

Comment:
I'm wondering why an error is being thrown when the readAll operation fails while using the getXXXX() API. Isn't the failure supposed to be handled by readAll, with the API's caller responsible for managing any failure actions? Shouldn't the getXXXX() API just return the last updated value in such cases?

I'm also thinking aloud: how would the caller of getXXXX() know how to handle the situation if the get call fails?

Reply:
readAll() is just going to be called every N seconds.  If there is a failure, none of the values are available until everything can be read again.  The design wants it to be all or nothing, so there's no way code can use a mix of new values along with stale or unavailable.